docs/key-issues/multiple_architectures.md

```diff
@@ -0,0 +1,129 @@
+responsibility of the user to find (or compile) a binary that was compiled for
+their host CPU architecture.
+
+However, on occasion, we see an operating system platform where multiple CPU
+architectures are supported:
+
+* In the early days of Windows NT, both x86 and DEC Alpha CPUs were supported
+* Although Linux started as an x86 project, the Linux kernel is now available a
+  wide range of other CPU architectures, including ARM64, RISC-V, PowerPC, s390
+  and more.
+* Apple transitioned Mac hardware from PowerPC to Intel (x86-64) CPUs, providing
+  a forwards compatibility path for binaries
+* Apple is currently transitioning Mac hardware from Intel (x86-64) to
+  Apple Silicon (ARM64) CPUs, again providing a forwards compatibility
+  path
+* Apple supports ARMv6, ARMv7, ARMv7s, ARM64 and ARM64e on iOS
+* Android currently supports ARMv7, ARM64, x86, and x86-64; it has historically
+  also supported ARMv5 and MIPS
+
+CPU architecture compatibility is a necessary, but not sufficient criterion for
+determining binary compatibility. Even if two binaries are compiled for the same
+CPU architecture, that doesn't guarantee [ABI compatibility](abi.md).
+
+In some respects, CPU architecture compatibility could be considered a superset
+of [GPU compatibility](gpus.md). When dealing with multiple CPU architectures,
+there may be some overal with the solutions that can be used to support GPUs in
+native binaries.
+
+## Platform approaches for dealing with multiple architectures
+
+Three approaches have emerged for handling multiple CPU architectures.
+
+### Multiple binaries
+
+The minimal solution is to distribute multiple binaries. This is the approach
+that was used by Windows NT, and is currently supported by Linux. At time of
+distribution, an installer or other downloadable artefact is provided for each
+supported platform, and it is up to the user to select and download the correct
+artefact.
+
+### Archiving
+
+The approach taken by Android is very similar to the multiple binary approach,
+with some affordances and tooling to simplify distribution.
+
+When building an Android project, each target architecture is compiled
+independently. If a native binary library is required to compile the Android
+application, a version must be provided for each supported CPU architecture. A
+directory layout convention exists for providing a binary for each platform,
+with the same library name. This yields an independent final binary (APK) for
+each CPU architecture. When running locally, a CPU-specific APK will be
+uploaded to the simulator or test device.
+
+To simplify the process of distributing the application, at time of publication,
+a single Android App Bundle (AAB) is generated from the multiple CPU-specific
+APKs. This AAB contains binaries for all platforms that can be uploaded to an
+app store.
+
+When an end-user requests the installation of an app, the app store strips out the
+binary that is appropriate for the end-user's device.
```

Comment:
In the Android case, it's not really a *problem* per se; its more of a "consideration to keep in mind".

Let's say I have an Android project that uses a Python package A that has a binary module; that package has a dependency on B (which also has a binary module). When I do the compilation pass for ARM64, I `pip install A`, which does a dependency resolution pass to determine a compatible version of B. I then do a compilation pass for ARMv7, which does a separate dependency resolution pass to determine a compatible version of B. The use of a separate dependency resolution pass introduces the possibility that my ARMv7 and ARM64 binaries have entirely different versions of package B (and, I guess, potentially package A as well). 

That's not a problem; it could even be argued as "working as designed". It's how BeeWare (strictly, the Chaquopy subsystem that Briefcase uses) is working at present; it just might be a little surprising if you don't have a systematic way of at least flagging that there are different versions installed for each platform. I'll modify the wording to clarify this.

However, for the same project on iOS, this isn't an option. iOS apps perform one compilation pass per ABI (i.e. one for the simulator, and one for the physical device), so the final installed artefact needs to contain a single fat binary with all the architectures. This *could* be treated as an install-time problem rather than distribution problem, though; I'll add a note about that possible approach.

Reply:
Thanks, this is a good and concrete example. It sounds like it's only a problem with `pip`/PyPI - any other package relevant package manager gets all the metadata first and then does a single solve for all dependencies, so you'd get a single version of package B.

Reply:
I'm not sure I follow why it's only a pip/PyPI problem. The issue isn't the solution path; it's how many times you need to run the installer (and thus the solver). If I solve and install for arm64, then solve and install for x86-64, that's 2 independent solutions, and any inconsistency in availability of packages for arm64 and x86-64 will result in different solutions. AFAICT, this will still be an issue for conda, as the issue isn't the availability of metadata for a single platform; it's the availability of the solution arrived at by a completely independent solver pass. 

You can only avoid this problem if you do a single resolver pass looking to satisfy *both* architectures at the same time, and only install once a single solution that works for both architectures is found (or, I guess, you have some ways to pass a specific solution found by pass 1 into subsequent passes in a way that *doesn't* include package hashes, as you'll be installing a different package with the same name and version, but different ABI/architecture).

Reply:
Ah okay, thank makes sense, thanks.

Reply:
> as the issue isn't the availability of metadata for a single platform; it's the availability of the solution arrived at by a completely independent solver pass.

conda metadata is available offline, unlike pip where every resolver run is a series of API calls followed by downloading artifacts followed by extracting artifacts followed by yet more API calls. So you actually *can* guarantee multiple runs of the resolver will produce the same results.

If the metadata is guaranteed to be consistent across architectures at any given moment (e.g. packages of any given version will always exist for all architectures) then you can also get the same resolver result across architectures too.

> or, I guess, you have some ways to pass a specific solution found by pass 1 into subsequent passes in a way that _doesn't_ include package hashes

pip freeze?

Reply:
> So you actually _can_ guarantee multiple runs of the resolver will produce the same results.

In principle, yes. But I don't think either `conda` or `mamba` have this capability, so in practice it'd be pretty hard today.